File: packages/wasm-dpp/test/unit/identity/stateTransition/IdentityUpdateTransition/IdentityUpdateTransition.spec.js
const getIdentityUpdateTransitionFixture = require('../../../../../lib/test/fixtures/getIdentityUpdateTransitionFixture');

const { IdentityPublicKey, Identifier, IdentityPublicKeyWithWitness } = require('../../../../..');
const { StateTransitionTypes } = require('../../../../..');
const generateRandomIdentifierAsync = require('../../../../../lib/test/utils/generateRandomIdentifierAsync');

describe('IdentityUpdateTransition', () => {
  let rawStateTransition;
  let stateTransition;

  beforeEach(async () => {
    stateTransition = await getIdentityUpdateTransitionFixture();
    rawStateTransition = stateTransition.toObject();
  });

  describe('#getType', () => {
    it('should return IDENTITY_UPDATE type', () => {
      expect(stateTransition.getType()).to.equal(StateTransitionTypes.IdentityUpdate);
    });
  });

  describe('#setIdentityId', () => {
    it('should set identityId', async () => {
      const id = await generateRandomIdentifierAsync();

      stateTransition.setIdentityId(id);

      expect(stateTransition.identityId.toBuffer())
        .to.deep.equal(id.toBuffer());
    });
  });

  describe('#getIdentityId', () => {
    it('should return identityId', () => {
      expect(stateTransition.getIdentityId().toBuffer())
        .to.deep.equal(rawStateTransition.identityId);
    });
  });

  describe('#getRevision', () => {
    it('should return revision', () => {
      expect(stateTransition.getRevision()).to.equal(rawStateTransition.revision);
    });
  });

  describe('#setRevision', () => {
    it('should set revision', () => {
      stateTransition.setRevision(42);

      expect(stateTransition.getRevision()).to.equal(42);
    });
  });

  describe('#getOwnerId', () => {
    it('should return owner id', () => {
      expect(stateTransition.getOwnerId().toBuffer()).to.deep.equal(
        rawStateTransition.identityId,
      );
    });
  });

  describe('#getPublicKeysToAdd', () => {
    it('should return public keys to add', () => {
      expect(stateTransition.getPublicKeysToAdd().map((key) => key.toObject()))
        .to.deep.equal(
          rawStateTransition.addPublicKeys,
        );
    });
  });

  describe('#setPublicKeysToAdd', () => {
    it('should set public keys to add', () => {
      const key = new IdentityPublicKeyWithWitness(1);
      key.setType(IdentityPublicKey.TYPES.BLS12_381);
      key.setData(Buffer.from('01fac99ca2c8f39c286717c213e190aba4b7af76db320ec43f479b7d9a2012313a0ae59ca576edf801444bc694686694', 'hex'));

      const publicKeys = [key];

      stateTransition.setPublicKeysToAdd(publicKeys);

      expect(stateTransition.addPublicKeys.map((k) => k.toObject()))
        .to.have.deep.members(publicKeys.map((k) => k.toObject()));
    });
  });

  describe('#getPublicKeyIdsToDisable', () => {
    it('should return public key ids to disable', () => {
      expect(stateTransition.getPublicKeyIdsToDisable())
        .to.deep.equal(rawStateTransition.disablePublicKeys);
    });
  });

  describe('#setPublicKeyIdsToDisable', () => {
    it('should set public key ids to disable', () => {
      stateTransition.setPublicKeyIdsToDisable([1, 2]);

      expect(stateTransition.getPublicKeyIdsToDisable())
        .to.deep.equal([1, 2]);
    });
  });

  describe('#toObject', () => {
    it('should return raw state transition', () => {
      rawStateTransition = stateTransition.toObject();

      expect(rawStateTransition).to.deep.equal({
        $version: '0',
        type: StateTransitionTypes.IdentityUpdate,
        signature: undefined,
        identityId: rawStateTransition.identityId,
        revision: rawStateTransition.revision,
        addPublicKeys: rawStateTransition.addPublicKeys,
        disablePublicKeys: rawStateTransition.disablePublicKeys,
        signaturePublicKeyId: 0,
      });
    });

    it('should return raw state transition without signature', () => {
      rawStateTransition = stateTransition.toObject({ skipSignature: true });

      expect(rawStateTransition).to.deep.equal({
        $version: '0',
        type: StateTransitionTypes.IdentityUpdate,
        identityId: rawStateTransition.identityId,
        revision: rawStateTransition.revision,
        addPublicKeys: rawStateTransition.addPublicKeys,
        disablePublicKeys: rawStateTransition.disablePublicKeys,
      });
    });

    it('should return raw state transition without optional properties', () => {
      stateTransition.setPublicKeyIdsToDisable(undefined);
      stateTransition.setPublicKeysToAdd(undefined);

      rawStateTransition = stateTransition.toObject();

      expect(rawStateTransition).to.deep.equal({
        $version: '0',
        type: StateTransitionTypes.IdentityUpdate,
        signature: undefined,
        identityId: rawStateTransition.identityId,
        revision: rawStateTransition.revision,
        signaturePublicKeyId: 0,
      });
    });
  });

  describe('#toJSON ', () => {
    it('should return JSON representation of state transition', () => {
      const jsonStateTransition = stateTransition.toJSON();

      expect(jsonStateTransition).to.deep.equal({
        $version: '0',
        type: StateTransitionTypes.IdentityUpdate,
        signature: undefined,
        identityId: stateTransition.getIdentityId().toString(),
        revision: rawStateTransition.revision,
        addPublicKeys: stateTransition.getPublicKeysToAdd().map((k) => k.toJSON()),
        disablePublicKeys: rawStateTransition.disablePublicKeys,
        signaturePublicKeyId: 0,
      });
    });
  });

  describe('#getModifiedDataIds', () => {
    it('should return ids of topped up identity', () => {
      const result = stateTransition.getModifiedDataIds();

      expect(result.length).to.be.equal(1);
      const identityId = result[0];

      expect(identityId).to.be.an.instanceOf(Identifier);
      expect(identityId.toBuffer()).to.be.deep.equal(rawStateTransition.identityId);
    });
  });

  describe('#isDataContractStateTransition', () => {
    it('should return false', () => {
      expect(stateTransition.isDataContractStateTransition()).to.be.false();
    });
  });

  describe('#isDocumentStateTransition', () => {
    it('should return false', () => {
      expect(stateTransition.isDocumentStateTransition()).to.be.false();
    });
  });

  describe('#isIdentityStateTransition', () => {
    it('should return true', () => {
      expect(stateTransition.isIdentityStateTransition()).to.be.true();
    });
  });
});

File: packages/rs-dpp/src/old/tests/identity/state_transition/identity_update_transition/identity_update_transition_spec.rs
use chrono::Utc;
use platform_value::{platform_value, BinaryData, Value};

use crate::identity::state_transition::identity_public_key_transitions::IdentityPublicKeyInCreation;
use crate::prelude::Revision;
use crate::{
    identity::{
        state_transition::identity_update_transition::identity_update_transition::IdentityUpdateTransition,
        KeyType, Purpose, SecurityLevel,
    },
    state_transition::{
        StateTransitionFieldTypes, StateTransitionIdentitySignedV0, StateTransitionType,
    },
    tests::{
        fixtures::get_identity_update_transition_fixture, utils::generate_random_identifier_struct,
    },
};

struct TestData {
    transition: IdentityUpdateTransition,
    raw_transition: Value,
}

fn setup_test() -> TestData {
    let transition = get_identity_update_transition_fixture();
    let raw_transition = transition.to_object(false).unwrap();
    TestData {
        transition,
        raw_transition,
    }
}

#[test]
fn get_type() {
    let TestData { transition, .. } = setup_test();
    assert_eq!(StateTransitionType::IdentityUpdate, transition.get_type());
}

#[test]
fn set_identity_id() {
    let TestData { mut transition, .. } = setup_test();
    let id = generate_random_identifier_struct();
    transition.set_identity_id(id);
    assert_eq!(&id, transition.get_identity_id());
}

#[test]
fn get_revision() {
    let TestData { transition, .. } = setup_test();
    assert_eq!(0, transition.get_revision());
}

#[test]
fn set_revision() {
    let TestData { mut transition, .. } = setup_test();
    transition.set_revision(42);
    assert_eq!(42, transition.get_revision());
}

#[test]
fn get_owner_id() {
    let TestData { transition, .. } = setup_test();
    assert_eq!(&transition.identity_id, transition.get_owner_id());
}

#[test]
fn get_public_keys_to_add() {
    let TestData { transition, .. } = setup_test();
    assert_eq!(
        &transition.add_public_keys,
        transition.get_public_keys_to_add()
    );
}

#[test]
fn set_public_keys_to_add() {
    let TestData { mut transition, .. } = setup_test();

    let id_public_key = IdentityPublicKeyInCreation {
        id: 0,
        key_type: KeyType::BLS12_381,
        purpose: Purpose::AUTHENTICATION,
        security_level : SecurityLevel::CRITICAL,
        read_only: true,
        data: BinaryData::new(hex::decode("01fac99ca2c8f39c286717c213e190aba4b7af76db320ec43f479b7d9a2012313a0ae59ca576edf801444bc694686694").unwrap()),
        signature : Default::default(),
    };

    transition.set_public_keys_to_add(vec![id_public_key.clone()]);

    assert_eq!(vec![id_public_key], transition.get_public_keys_to_add());
}

#[test]
fn get_disable_public_keys() {
    let TestData { transition, .. } = setup_test();
    assert_eq!(
        transition.disable_public_keys,
        transition.get_public_key_ids_to_disable()
    );
}

#[test]
fn set_disable_public_keys() {
    let TestData { mut transition, .. } = setup_test();
    let id_to_disable = vec![1, 2];
    transition.set_public_key_ids_to_disable(id_to_disable.clone());

    assert_eq!(&id_to_disable, transition.get_public_key_ids_to_disable());
}

#[test]
fn get_public_key_disabled_at() {
    let TestData { transition, .. } = setup_test();
    assert_eq!(
        transition.public_keys_disabled_at,
        transition.get_public_keys_disabled_at()
    );
}

#[test]
fn set_public_key_disabled_at() {
    let TestData { mut transition, .. } = setup_test();
    let now = Utc::now().timestamp_millis() as u64;
    transition.set_public_keys_disabled_at(Some(now));

    assert_eq!(Some(now), transition.get_public_keys_disabled_at());
}

#[test]
fn to_object() {
    let TestData { transition, .. } = setup_test();
    let result = transition
        .to_object(false)
        .expect("conversion to object shouldn't fail");

    let expected_raw_state_transition = platform_value!({
        "protocolVersion" : 1u32,
        "type" : 5u8,

        "identityId" : transition.identity_id,
        "revision": 0 as Revision,
        "addPublicKeys" : [
            {

                "id" : 3u32,
                "type": 0u8,
                "purpose" : 0u8,
                "securityLevel" : 0u8,
                "readOnly" : false,
                "data" :BinaryData::new(base64::decode("AkVuTKyF3YgKLAQlLEtaUL2HTditwGILfWUVqjzYnIgH").unwrap()),
                "signature" : BinaryData::new(vec![0u8;65])
            }
        ],
        "disablePublicKeys" : [0u32],
        "publicKeysDisabledAt" : 1234567u64,
        "signaturePublicKeyId": 0u32,
        "signature" : BinaryData::new(vec![0u8;65]),
    });

    assert_eq!(expected_raw_state_transition, result);
}

#[test]
fn to_object_with_signature_skipped() {
    let TestData { transition, .. } = setup_test();
    let result = transition
        .to_object(true)
        .expect("conversion to object shouldn't fail");

    let expected_raw_state_transition = platform_value!({
        "protocolVersion" : 1u32,
        "type" : 5u8,
        "identityId" : transition.identity_id,
        "revision": 0 as Revision,
        "addPublicKeys" : [
            {

                "id" : 3u32,
                "type": 0u8,
                "purpose" : 0u8,
                "securityLevel" : 0u8,
                "readOnly" : false,
                "data" :BinaryData::new(base64::decode("AkVuTKyF3YgKLAQlLEtaUL2HTditwGILfWUVqjzYnIgH").unwrap()),
            }
        ],
        "disablePublicKeys" : [0u32],
        "publicKeysDisabledAt" : 1234567u64,
    });

    assert_eq!(expected_raw_state_transition, result);
}

#[test]
fn to_json() {
    let TestData { transition, .. } = setup_test();
    let result = transition
        .to_object(false)
        .expect("conversion to platform value shouldn't fail");

    let expected_raw_state_transition = platform_value!({
        "protocolVersion" : 1u32,
        "type" : 5u8,
        "identityId" : transition.identity_id,
        "revision": 0 as Revision,
        "addPublicKeys" : [
            {

                "id" : 3u32,
                "type": 0u8,
                "purpose" : 0u8,
                "securityLevel" : 0u8,
                "readOnly" : false,
                "data" : BinaryData::new(base64::decode("AkVuTKyF3YgKLAQlLEtaUL2HTditwGILfWUVqjzYnIgH").unwrap()),
                "signature" : BinaryData::new(vec![0;65]),
            }
        ],
        "disablePublicKeys" : [0u32],
        "publicKeysDisabledAt" : 1234567u64,
        "signaturePublicKeyId": 0u32,
        "signature" : BinaryData::new(vec![0u8;65]),
    });

    assert_eq!(expected_raw_state_transition, result);
}
